### PR TITLE
Provide remove_dir_contents

### PR DESCRIPTION
**Pull request by ijackson**
Hi.

Apropos of https://github.com/rust-lang/rustup/issues/2433 @rbtcollins suggested putting the `remove_dir_contents` function here in `remove_dir_all`.

I think I have done the right thing in the Windows code in `fs.rs` <strike>but I have not been able to test, or even compile, the Windows code at all because I have no Windows installs</strike>.  _Edit: the CI tests seem to pass so I think it is probably right..._

I did provide a test case for my new code.

I hope this meets with your approval.  Let me know what you think.

Regards,
Ian.

### CODE REVIEW

**Comment by ijackson** [comment]
Looks like I am going to debug this via the CI.  I hope this is not too annoying to you.

**Comment by ijackson** [comment]
Second time lucky I guess.  Out of interest are any of those CI tests on Unix?  Perhaps not, but if you accept this MR then maybe there should be since you will have nontrivial Unix and allegedly-portable code.

Anyway, thanks for your attention and I look forward to your review.

**Comment by XAMPPRocky** [comment]
Thank you for your PR! This PR looks good to me. Currently the CI only tests windows, and I think I'd want to have some testing on unix before merging this. If you'd like I'd accept adding a new job to the CI that tests unix in this PR? It should be fairly straightforward, otherwise I can write the CI config in a bit.

**Comment by ijackson** [comment]
XAMPPRocky writes ("Re: [XAMPPRocky/remove_dir_all] Provide remove_dir_contents (#22)"):
> Thank you for your PR! This PR looks good to me.

Thanks!

> Currently the CI only tests windows, and I think I'd want to have
> some testing on unix before merging this.  If you'd like I'd accept
> adding a new job to the CI that tests unix in this PR?  It should be
> fairly straightforward, otherwise I can write the CI config in a
> bit.

I haven't ever messed with github CI so if you know what you are doing
I would appreciate it if you would do it.

Regards,
Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

**Comment by XAMPPRocky** [comment]
FWIW GitHub actions is pretty easy to use especially if you use the GitHub online editor which has auto-completion. I've added the unix CI to master so you should just need to rebase the changes. https://github.com/XAMPPRocky/remove_dir_all/commit/775feb136844e66d8db8bfb7a3db05efb542c8e1

**Comment by ijackson** [comment]
XAMPPRocky writes ("Re: [XAMPPRocky/remove_dir_all] Provide remove_dir_contents (#22)"):
> FWIW GitHub actions is pretty easy to use especially if you use the GitHub
> online editor which has auto-completion. I've added the unix CI to master so
> you should just need to rebase the changes. 775feb1

Thanks.  I looked at what you did.  Frankly I think it would have
taken me a while :-), so thank you for that.  I may well use this as a
reference next time I need to add a relevant github test.

I force pushed my branch and the CI seems to have liked it.

Thanks for your help!

Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

File: src/portable.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+use std::fs;
+use std::io;
+use std::path::Path;
+
+/// Deletes the contents of `dir_path`, but not the directory iteself.
+///
+/// If `dir_path` is a symlink to a directory, deletes the contents
+/// of that directory.  If `dir_path` does not exist, simply returns
+/// (without deleting anything).
+pub fn remove_dir_contents(dir_path: &Path) -> Result<(), io::Error> {
```
Comment (rbtcollins) [inline src/portable.rs:10]: I can see why you've written this new function, but I'd like to suggest a different implementation - bearing in mind some future optimisations I have planned. In particular, I want to avoid this serial loop over the read_dir contents at the top layer.

My suggestion is that rather than having allow_nondirectory / directories_only, your new flag to the core workhorse is delete_root; and that the second entry point to it be remove_dir_contents; then the only change needed is to move the remove_item(path, ctx) call from remove_dir_all_recursive to the caller of it, which will allow it to be skipped for the root parent case.

e.g.

```
 ctx.readonly = child.metadata()?.permissions().readonly();
        if child_type.is_dir() {
            remove_dir_all_recursive(&child.path(), ctx)?;
+          remove_item(&child.path(), ctx)
        } else {
            remove_item(&child.path().as_ref(), ctx)?;
        }
    }
    ctx.readonly = dir_readonly;
-    remove_item(path, ctx)
```

And call it from the root case as well.

This of course leaves the posix case needing its own implementation, but I think it will be very shallow at that point.

**Comment by ijackson** [comment]
Robert Collins writes ("Re: [XAMPPRocky/remove_dir_all] Provide remove_dir_contents (#22)"):
> @rbtcollins commented on this pull request.
> 
> Thank you for doing this, and I'm sorry I'm so late getting to a review for
> you.

No problem.
> +fn remove_some<P: AsRef<Path>>(path: P, allow_nondirectory: bool) -> io::Result<()> {
> 
> I find allow_nondirectory a little hard to parse: is this !"directories_only" ?
> If so perhaps change this to directories_only or similar and invert the sense
> in the code below.

Sure (subject to changes discussed below...)

> In src/portable.rs:
> 
> > @@ -0,0 +1,70 @@
> +use std::fs;
> +use std::io;
> +use std::path::Path;
> +
> +/// Deletes the contents of `dir_path`, but not the directory iteself.
> +///
> +/// If `dir_path` is a symlink to a directory, deletes the contents
> +/// of that directory.  If `dir_path` does not exist, simply returns
> 
> I think if dir_path doesn't exist, this should return ENOENT - well,
> rust style - e.g. don't capture the io::ErrorKind::NotFound special
> case. For several reasons: we don't special case that in the
> remove_dir() case; calling code that wants to (for instance) ensure
> an empty dir will be more complex if this error is hidden, and its
> not needed for the use case we're adding this for.

Unfortunately doing that will make it impossible to distinguish "the
top level directory did not exist" from "during the innards of
attempting to recursively delete something we unexpectedly got
ENOENT".  I expect the current code just bubbles that unexpected
ENOENT up to its caller (although I haven't checked).

I have a counterproposal: leave this entrypoint as I had it, but
also provide:

  /// Ensures that `dir_path` is an empty directory, or a symlink to
  /// one.  If it already exists, its contents are deleted (as if
  /// with `remove_dir_contents`.  If it does not, it is created
  /// (as if with `std::fs:create_dir`.)
  fn ensure_empty_dir(...)

This is surely what many callers (including rustup) actually want.
I think the obvious implementation is mkdir, followed if necessary by
remove_dir_contents.

> In src/portable.rs:
...
> I can see why you've written this new function, but I'd like to suggest a
> different implementation - bearing in mind some future optimisations I have
> planned. In particular, I want to avoid this serial loop over the read_dir
> contents at the top layer.
> 
> My suggestion is that rather than having allow_nondirectory /
> directories_only, your new flag to the core workhorse is
> delete_root; and that the second entry point to it be
> remove_dir_contents; then the only change needed is to move the
> remove_item(path, ctx) call from remove_dir_all_recursive to the
> caller of it, which will allow it to be skipped for the root parent
> case.

I'll have look at the code.

> This of course leaves the posix case needing its own implementation, but I
> think it will be very shallow at that point.

The POSIX case can just do the loop, I think ?

Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

**Comment by rbtcollins** [comment]
> Unfortunately doing that will make it impossible to distinguish "the top level directory did not exist" from "during the innards of attempting to recursively delete something we unexpectedly got ENOENT". I expect the current code just bubbles that unexpected ENOENT up to its caller (although I haven't checked). I have a counterproposal: leave this entrypoint as I had it, but also provide: 

The current code also doesn't discriminate between ENOENT of the top level not existing and an ENOENT of a file that is being deleted - and actually thats a bug; it should handle ENOENT of a file being deleted by ignoring that error - it simply indicates that the job it is tasked with has been made easier, after all. There is some handling of related cases with hard links in fact,

So while I agree with the defect today, I think we should fix it, which would remove the issue.

/// Ensures that `dir_path` is an empty directory, or a symlink to 
/// one. If it already exists, its contents are deleted (as if 
/// with `remove_dir_contents`. If it does not, it is created 
/// (as if with `std::fs:create_dir`.) fn ensure_empty_dir(...) This is surely what many callers
/// (including rustup) actually want. I think the obvious implementation is mkdir, followed
/// if necessary by remove_dir_contents.

I think that that would be nice to offer as well.

> In src/portable.rs:
> ...
> I can see why you've written this new function, but I'd like to suggest a different implementation - bearing in mind some future optimisations I have planned. In particular, I want to avoid this serial loop over the read_dir contents at the top layer. My suggestion is that rather than having allow_nondirectory / directories_only, your new flag to the core workhorse is delete_root; and that the second entry point to it be remove_dir_contents; then the only change needed is to move the remove_item(path, ctx) call from remove_dir_all_recursive to the caller of it, which will allow it to be skipped for the root parent case.
> I'll have look at the code.
> This of course leaves the posix case needing its own implementation, but I think it will be very shallow at that point.
> The POSIX case can just do the loop, I think ? Ian.

Yes; the IO latency behaviour of POSIX platforms is generally much roundabout than Windows.

-Rob

File: src/portable.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
+use std::fs;
+use std::io;
+use std::path::Path;
+
+/// Deletes the contents of `dir_path`, but not the directory iteself.
+///
+/// If `dir_path` is a symlink to a directory, deletes the contents
+/// of that directory.  If `dir_path` does not exist, simply returns
```
[the remark this answers is not in the document]
Comment (ijackson) [reply src/portable.rs]: I have changed this in my new version.

File: src/portable.rs (hunk 1 above)
Comment (ijackson) [reply src/portable.rs]: I looked at master, and yes, a different implementation strategy was sensible.  I found it better to split the function rather than using a boolean flag - I prefer to encode things in control flow rather than flag booleans, as that makes analysis (eg type checks) easier.  I hope you like what I have done.

**Comment by ijackson** [comment]
I'm afraid I am going to have to iterate via the CI since I do not have access to any Windows machines.  Please bear with me.

**Comment by rbtcollins** [comment]
All good. By the way I'll probably be making the threaded implementation an
opt in feature, as tempdir objected to it (which is a longer discussion
than fits here, but suffice to say that windows is not POSIX and the norms
and expectations are a source of friction)

**Comment by ijackson** [comment]
People complain about rustc being slow but they should try the compiler error whack-a-mole this way!

**Comment by ijackson** [comment]
I don't think I want to think about tempdir and Windows and multithreading....

Anyway, looks like this might be going to pass its checks now.  Thanks for the review etc.

**Comment by rbtcollins** [comment]
If you want a faster means, there are free beer VM images of windows
available for use for developers, such as you.

**Comment by ijackson** [comment]
I'm hoping this will be a rare problem for me :-).

**Comment by ijackson** [comment]
Hi.  I hope this is ready to merge now?

**Comment by rbtcollins** [comment]
I haven't had time to review sorry; moving across the world consumed all my recent bandwidth. Just coming out of that now.

File: README.md [hunk 3]
```diff
@@ -9,6 +9,12 @@
 A reliable implementation of `remove_dir_all` for Windows. For Unix systems
 re-exports `std::fs::remove_dir_all`.
 
+Also provides `remove_dir_contents` for both Windows and Unix.  This
+removes just the contents of the directory, if it exists.
+
+And provides `ensure_e pty_dir` for both Windows and Unix.  This
```
Comment (rbtcollins) [inline README.md:15]: There's a typo here.

File: src/portable.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+pub fn remove_dir_contents<P: AsRef<Path>>(path: P) -> io::Result<()> {
+    // This wrapper function exists because the core function
+    // for Windows, in crate::fs, returns a PathBuf, which our
+    // caller shouldn't see.
+    _remove_dir_contents(path)?;
+    Ok(())
+}
+
+/// Makes `path` an empty directory: if it does not exist, it is
+/// created it as an empty directory (as if with
```
Comment (rbtcollins) [inline src/portable.rs:23]: There's some grammar issue here - "it is created it as an"

File: src/portable.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+/// `std::fs::create_dir`); if it does exist, its contents are
+/// deleted (as if with `remove_dir_contents`).
+///
+/// It is an error of `path` exists but is not a directory (or
```
Comment (rbtcollins) [inline src/portable.rs:27]: of instead of if?

**Comment by rbtcollins** [comment]
It's certainly a lot cleaner now. Which of the new APIs do you anticipate rustup using? Do we actually need the other one? ensure_empty_dir kindof implies the existence of an ensure_empty_dir_all call as well perhaps, though if people are deleting all of ~/.rustup and expecting rustup to be ok with that .. well, thats a different discussion.

**Comment by ijackson** [comment]
Thanks for the review.  I'll fix the typos and grammar.

> It's certainly a lot cleaner now. Which of the new APIs do you anticipate rustup using?

I was imaginging replacing the deletion call at the end of rustup, with remove_dir_contents, and not touching rustup's directory creation code.  Correct me if I'm wrong, but presumably rustup already does not mind finding leftover stuff from a previous run in its temporary directories.  Otherwise it would sometimes get into a stuck state.

> Do we actually need the other one? ensure_empty_dir kind of implies the existence of an ensure_empty_dir_all

I think something like ensure_empty_dir or ensure_empty_dir_all would be extremely useful for tests, which often want to make working files, but want to leave the test artefacts lying about for diagnosis.  (There is a can of worms here for rust tests because of cargo's failure to provide a formal place to put such things, but let's leave that for anoyther day.)

If you prefer not to take the new ensure_empty_dir call now then that's fine by me.  But it seems that if someone wanted such a thing in their own project, I might well think this crate a good place for it to live.  The same applies to ensure_empty_dir_all; so, alternatively, if you want me to write that one too I'm happy to do that.

**Comment by ijackson** [comment]
Hi.  Did you have an opinion about ensure_empty_dir etc. ?

**Comment by ijackson** [comment]
@rbtcollins Hi, what would you like to do here?  Thanks, Ian.

**Comment by rbtcollins** [comment]
Sorry, this is on my todo, but I'm timeslicing like mad right now, please
leave with me for the moment.

-Rob

File: src/portable.rs (hunk 5 above)
Comment (J-Vin) [reply src/portable.rs:27]: if

**Review by J-Vin — APPROVED** [review]

**Comment by ijackson** [comment]
@J-Vin thanks for the review :-).  I fixed that typo and rebased onto current main.

**Comment by rbtcollins** [comment]
Sorry about how long this took me to get to - hazards of moving across the world.
